### PR TITLE
backport of #4761, #4771 to v1.8.x

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
backport of #4761, #4771

## Why ?
don't block CI for other PRs

### CODE REVIEW

**Comment by evgeny-leksikov** [comment]
rnat
bot:mlx:retest

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 2 of 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 12-Mar-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/cd08b2a87c56690398dfb05314749ae0f9c71505/_main) | :question: ABORTED
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/7cbac406bc73bd1febba196580e26de2237f5f88/hpc-arm-hwi-jenkins_W1) | :question: ABORTED
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/5aa3e6946d903ccd47a66edcf415d5d3f6de83d3/hpc-test-node-gpu_W0) | :question: ABORTED
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/c317daf4b6a8c8131ac486b627104e481e403c0c/hpc-test-node-gpu_W1) | :x: FAILURE
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/c345c20ab126a56e5571dcd2a352c5e0b9e2d06e/hpc-test-node-gpu_W3) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/eac0a0a5d710b8f0bff4d1ac9661e5d8182454c8/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/1a91c5864da29d1f4b40ccea0c0de7fc099edff7/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/52758fbcd401814fe5f717cb285d6e91736ae358/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/735483d41dd2df491cdc086dccd35da9ec4dc5c4/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/ba00f1a00ea38cf69a16b25260ebf91138dab2ad/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/4cb25cbdc12bcd4fb229c283ea28251cc380e631/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/ba3b08d343ee6f4d77aee7fcc4da580a7f773b19/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/7d7319e38c6d7bd00dab84cd1b6899772e97de5c/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/c4ad9d6d15bd6b9f395954584316870efbb7850c/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/7914543033a8864ee6ecf0792189a6108c35d0a1/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/ca5afca8a7adbbd987465a337ebc14357c452c2f/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/4b3d490917dd7d07cd2ab5b7104fe75bb7cce61d/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/d5c4fa527225281a2a58de363926ea0fbe9a0eb8/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/1c3c05f1aa42f05351d2cba4674a9b48a971f7d9/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/7570633b2fe40a8f1cb20b9820791435ab72c6f4/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/0167c1e5b0522769c0e07318433c7ae5ecfce96d/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/d847672920721b31152ffde4e40ea8a0beea656c/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/2d6ca450b9fc9d80f93cbce3f95cbf1c601d5a01/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/90f5b329c9ec6b15be45628eebda4a975b751278/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/931e5ec423fdb7cbd8c4bae3eb9949ca/raw/0ca7f033902357949c2d5afdf00dce19a405f03d/hpc-test-node-gpu_W2) | :question: UNKNOWN

**Comment by yosefe** [comment]
#4798 :(
bot:mlx:retest

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 13-Mar-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/410e6ec71296ea94c1c0b8f5241bad2ef2b94821/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/287a601a22a6aeb8a3a61277d19a4a36b04d4cc2/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/97a5091a3a1ca4b65a13fc1e9449feef045d0a59/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/a91f527da707ea014f075cba122476c091573959/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/e1cde617c03093f6fda7f7c2cdec7c0263e90aff/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/03a8d75c512a5045e4ded9631033d8c704930896/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/52df6deb1bd37980a6734c95f585c2a3b3eae13d/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/aa87c67fce6cb1f042846f855b1991190b940887/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/5f95c9ccf8d1e99ab827f6c05240e48bad5d6d4b/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/5c36b32374b234f5b4c1ded1351e31659e110309/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/c863a3f6e61bcd2a737cc0eda3fdeb73a26debf0/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/48c1077225c7731155e7161a02a0ff53d9fa1047/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/609b174ee89120cfbd31caf751fe91e781ac808b/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/640bfdc51238cd4fa5c7e504f69d93d6725e6922/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/fa5565d4de2b76ab03281721ded8c4cef2f48a3d/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/69352ce5566844624cf778291f92e9fa53682807/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/958699cda3e4859ffee1391949073cbe84095b84/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/d7760936e3252b2c3726f64d34eea343ad430912/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/9771f976582b0ef949dcc84ec7984a1e40f63293/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/3f51d2d79d3c27019443157ae30d96abb5439c69/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/ca93c33eea721bf15a558567cdda90457cf8e1f7/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/023dda23da028f13a312433d28f2602022ab4e91/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/1ef379941a5513b41984e5003fa7e5b6b1cefb9a/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/3c7723f131c0a274d96293668cde347bedb95448/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/de889fd78cfe1b6f82ee2fa182bd169b/raw/6e57b3a4915cc3cc1c1de9b0255e7707ac7bc0c1/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

File: test/gtest/uct/uct_test.cc [hunk 1]
```diff
@@ -1252,11 +1252,11 @@ void uct_test::entity::listen(const ucs::sock_addr_storage &listen_addr,
         if (ifa_addr->sa_family == AF_INET) {
             struct sockaddr_in *addr =
                             reinterpret_cast<struct sockaddr_in *>(ifa_addr);
-            addr->sin_port = ucs::get_port();
+            addr->sin_port = ntohs(ucs::get_port());
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: can we use sock_addr_storage::get_port() ?

File: test/gtest/uct/uct_test.cc [hunk 2]
```diff
@@ -741,11 +741,11 @@ uct_test::entity::entity(const resource& resource, uct_iface_config_t *iface_con
         if (ifa_addr->sa_family == AF_INET) {
             struct sockaddr_in *addr =
                 reinterpret_cast<struct sockaddr_in *>(ifa_addr);
-            addr->sin_port = ucs::get_port();
+            addr->sin_port = ntohs(ucs::get_port());
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: can we use sock_addr_storage::get_port() ?

File: src/ucp/wireup/wireup_cm.c [hunk 3]
```diff
@@ -733,21 +737,33 @@ static void ucp_cm_server_connect_cb(uct_ep_h ep, void *arg,
...
+         * 1) establish connection to complete any pending requests from wireup
+         *    lane
+         * 2) handle disconnect same way as close protocol
+         * 3) TODO: remove (1) when the EP can be moved to err state to block
+         *          new send operations but still able to flush transport lanes */
+        uct_worker_progress_register_safe(ucp_ep->worker->uct,
+                                          ucp_cm_server_connect_progress,
+                                          ucp_ep, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                          &prog_id);
+        ucp_cm_disconnect_cb(ep, ucp_ep);
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c]: @yosefe `ucp_cm_disconnect_cb` adds another slow progress callback and calls `ucp_worker_signal_internal`, i don't see any reason to call `ucp_worker_signal_internal` twice.

File: test/gtest/uct/uct_test.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/gtest/uct/uct_test.cc]: `sock_addr_storage::get_port` does different thing, it just returns already initialized port number. 
we can add static function to `sock_addr_storage` like `get_free_port` if you think it makes sense.

Comment (yosefe) [reply test/gtest/uct/uct_test.cc]: @evgeny-leksikov sorry, i've meant set_port

File: src/ucp/wireup/wireup_cm.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c]: ok, how about uct_worker_progress_register_safe() call which looks the same in both cases?

[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c]: actually, i notices we are doing connect and then immediately disconnect. maybe we can short-circuit here?

[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c]: > how about uct_worker_progress_register_safe() call which looks the same in both cases?

I wanted to isolate logically different cases to make code cleaner. Moreover, one of them (RESET case) will be removed as described in `3) TODO: ...`

> maybe we can short-circuit here?

that is written in `3) TODO: ... `

File: test/gtest/uct/uct_test.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/gtest/uct/uct_test.cc]: yes, will PR to master

File: src/ucp/wireup/wireup_cm.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c]: ok then

File: test/gtest/uct/uct_test.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/gtest/uct/uct_test.cc]: @yosefe this is `ucs_sock_addr_t uct_test::entity::m_iface_params::mode::sockaddr::listen_sockaddr`, not a `sock_addr_storage`, so seems like it makes sense to keep `iface_params` as a solid field of `uct_test` for now

**Pull request merged**